plugin/plugin.go

```diff
@@ -363,6 +373,231 @@ func RegisterShutdown(name string, s Shutter) error {
+		assignSlice(oconfig, target)
+	} else if (oconfig.values_num == 1) && (isRoot == false) { // single value
+		assignValue(oconfig, target, 0)
+	} else if oconfig.children_num > 0 { // a container of oconfig children
+		for _, child := range getOconfigChildren(oconfig) {
+			childkey := C.GoString(child.key)
+			if target.FieldByName(childkey).IsValid() != true { // if there's no corresponding target field, skip it
+				fmt.Printf("Ignoring unexpected key in config: %v\n", childkey)
+				continue
+			}
+			if err := assignConfig(child, target.FieldByName(childkey), false); err != nil {
+				return fmt.Errorf("in %v: %v\n", childkey, err)
+			}
+		}
+	}
+	return nil
+}
+
+// RequestConfiguration registers a Configuration struct with the daemon and
+// requests a callback to fill it in. It returns a channel which will be
+// closed when a configuration has been loaded successfully
+func RequestConfiguration(name string, c Configuration) (chan struct{}, error) {
+	if err := checkUsable(reflect.ValueOf(c).Elem()); err != nil {
+		fmt.Printf("while loading plugin received error: %v\n", err)
+		return nil, fmt.Errorf("plugin contains unusable configuration struct")
+	}
+	if len(config_targets) == 0 {
+		config_targets = make(map[string]*Configuration)
+		configured = make(map[string]chan struct{})
+	}
+	config_targets[name] = &c
+	configured[name] = make(chan struct{})
+	cName := C.CString(name)
+	cCallback := C.plugin_complex_config_cb(C.process_complex_config)
+	status, err := C.register_complex_config_wrapper(cName, cCallback)
+	if err != nil {
+		Errorf("register_complex_config_wrapper failed with status %v", status)
+		return nil, err
+	}
+	return configured[name], nil
+}
+
+// process_complex_config receives the plugin config from Collectd and attempts to
+// initialise the target config. If the configuration is parsed without fatal errors
+// it calls the attached Validate() function, which allows the plugin developer to
+// confirm that they've received sufficient valid configuration. If that succeeds
+// we finally close the Configured channel to indicate that the configuration is now
+// assigned and validated.
+//export process_complex_config
+func process_complex_config(oconfig *C.oconfig_item_t) C.int {
+	oname := C.GoString(C.go_get_string_value(oconfig, C.int(0)))
+	var ok bool
+	if config_target, ok = config_targets[oname]; !ok {
+		fmt.Printf("Error: Received configuration for unknown plugin name %v\n", oname)
+		return C.int(1)
+	}
+	target := reflect.ValueOf(*config_target).Elem()
+	err := assignConfig(oconfig, target, true)
+	if err != nil {
+		fmt.Printf("Error: Invalid configuration %v\n", err)
+		return C.int(1)
+	}
+	valArgs := make([]reflect.Value, 0)
+	valResp := target.MethodByName("Validate").Call(valArgs)
+	if err := valResp[0].Interface(); err != nil {
+		fmt.Printf("Error: Plugin-specific validation returned error: %v\n", err)
+		return C.int(1)
+	}
+	close(configured[oname])
```

Comment:
Ouch. Yep, closing a channel is strictly a once-off - I hadn't realised we might see multiple calls for the same plugin name. 
I think the answer is to implement a handler for the initialisation callback and move both the `Validate()` and `close(configured[oname]` calls there.